docs/design-documents/hal/0001-i2c-overhaul.md

```diff
@@ -0,0 +1,527 @@
+- `i2c_free`:
+  - Resets the pins used to initialise the peripheral to their default state.
+  - Disables the peripheral clock.
+- `i2c_get_capabilities`:
+  - Fills the contents of the `i2c_capabilities_t` parameter
+- `i2c_frequency`:
+  - Returns the actual frequency that will be used.
+  - Sets the frequency to use for the transfer.
+  - Must leave all other configuration unchanged.
+- `i2c_timeout`:
+  - Sets the transmision timeout to use for the following blocking transfers.
+  - If the timeout is not set the default timeout is used.
+  - Default timeout value is based on I2C frequency. Is computed as triple amount of time it would take to send data over I2C
+- `i2c_write`:
+  - Writes `length` number of symbols to the bus.
+  - Returns the number of symbols sent to the bus.
+  - Returns an error code if transfer fails.
+  - Generates a stop condition on the bus at the end of the transfer if `stop` parameter is true.
+  - Handles transfer collisions and loss of arbitration if the platform supports multimaster in hardware.
+  - The transfer will timeout and return `I2C_ERROR_TIMEOUT ` if the transfer takes longer than the configured timeout duration.
+- `i2c_read`:
+  - Reads `length` symbols from the bus.
+  - Returns the number of symbols received from the bus.
+  - Returns an error code if transfer fails.
+  - Generates a stop condition on the bus at the end of the transfer if `stop` parameter is true.
+  - Handles transfer collisions and loss of arbitration if the platform supports multimaster in hardware.
+  - The transfer will timeout and return `I2C_ERROR_TIMEOUT ` if the transfer takes longer than the configured timeout duration.
+- `i2c_start`:
+  - Generates I2C START condition on the bus in master mode.
+  - Does nothing if called when the peripheral is configured in slave mode.
+- `i2c_stop`:
+  - Generates I2C STOP condition on the bus in master mode.
+  - Does nothing if called when the peripheral is configured in slave mode.
+- `i2c_slave_status`:
+  - Indicates which mode the peripheral has been addressed in.
+  - Returns not addressed when called in master mode.
+- `i2c_slave_address`:
+  - Sets the address of the peripheral to the `address` parameter.
+  - Does nothing if called in master mode.
+- `i2c_transfer_async`:
+  - Returns immediately with a `bool` indicating whether the transfer was successfully scheduled or not.
+  - The callback given to `i2c_transfer_async` is invoked when the transfer finishes or error occurs.
+  - Must save the handler and context pointers inside the `obj` pointer.
+  - The context pointer is passed to the callback on transfer completion.
+  - The callback must be invoked on completion unless the transfer is aborted.
+  - Handles transfer collisions and loss of arbitration if the platform supports multimaster in hardware.
+  - `i2c_async_event_t` must be filled with the number of symbols sent to the bus during transfer.
+- `i2c_abort_async`:
+  - Aborts any on-going async transfers.
```

Comment:
This should be optional - "best effort".

Reply:
changed